Python fixes, docs and restructure

kev-inn opened the pull request.

kev-inn committed:
[SYSTEMDS-312] Python 3.6 compatibility

Improvements to execution of building the python API.

Restructure python package.

Add initial documentation.

The install part will not fully work yet, because we did not yet put the package on pypi.

Remove `dag_node.rst`

Remove print in test case

Fix python workflow

I removed the `python setup.py build` call because the tests anyway
test the source files

Add some missing licenses

Fix rst comments

kev-inn commented on pom.xml:
```diff
@@ -1081,5 +1081,10 @@
+            <groupId>net.sf.py4j</groupId>
+            <artifactId>py4j</artifactId>
+            <version>0.10.9</version>
+        </dependency>
```
I do not know how we want to handle this during merge (0.2.0 prep release should anyway include py4j)

Baunsgaard replied:
It is indented using spaces, instead of tabs

kev-inn replied:
Oh yes, will quickly fix.

kev-inn commented on src/main/python/BUILD_INSTRUCTIONS.md:
````diff
@@ -0,0 +1,26 @@
+python3 create_python_dist.py
+```
+- now in the `./dist` directory there will exist the source distribution `systemds-VERSION.tar.gz` and the wheel distribution `systemds-VERSION-py3-none-any.whl`, with `VERSION` being the current version number
+- Finished. We can now upload it with `python3 -m twine upload --repository-url [URL] dist/*`
````
This can probably done simpler (have to actually distribute a package to check)

Baunsgaard replied:
seems fine.

kev-inn commented on src/main/python/LICENSE:
```diff
@@ -1,202 +0,0 @@
-                                 Apache License
```
This will be automatically be copied during `pre_setup.py`

kev-inn commented on src/main/python/NOTICE:
```diff
@@ -1,8 +0,0 @@
-SystemDS
```
This file will be automatically copied during `pre_setup.py`

kev-inn commented on src/main/python/README.md:
````diff
@@ -15,18 +15,3 @@ as well as distributed operations on Apache Spark. In contrast to existing syste
 provide homogeneous tensors or 2D Datasets - and in order to serve the entire
 data science lifecycle, the underlying data model are DataTensors, i.e.,
 tensors (multi-dimensional arrays) whose first dimension may have a heterogeneous and nested schema.
-
-## Build instructions
-
-- Cd to this folder.
-- Run pre_setup.py
-
-```bash
-python3 pre_setup.py
-```
-
-- Run setup.py
-
-```bash
-python3 setup.py build
-```
````
I decided to remove this, because this will show up on pypi. I instead created `BUILD_INSTRUCTIONS.md`

Baunsgaard replied:
:+1:

kev-inn commented on src/main/python/docs/source/simple_examples.rst:
```diff
@@ -0,0 +1,77 @@
+More complex operations
+-----------------------
+
+SystemDS provides high level functions for Data-Scientists, lets take a look at l2svm::
```
Could someone improve this sentence? I think @mboehm7 had a good phrase on his slides, but I don't remember it exactly.

kev-inn commented on src/main/python/pre_setup.py:
```diff
@@ -36,14 +36,12 @@
     if any((fnmatch.fnmatch(file, valid_name) for valid_name in JAR_FILE_NAMES)) and not any(
             (fnmatch.fnmatch(file, exclude_name) for exclude_name in EXCLUDE_JAR_FILE_NAMES)):
-        shutil.copyfile(os.path.join(root_dir, 'target', file), os.path.join(java_dir_full_path, 'lib', file))
-    if file == "lib":
-        shutil.copytree(os.path.join(root_dir, 'target', file), os.path.join(java_dir_full_path, 'lib', file))
-		
+        shutil.copyfile(os.path.join(root_dir, 'target', file), os.path.join(java_dir_full_path, file))
+    if file == 'lib':
+        shutil.copytree(os.path.join(root_dir, 'target', file), os.path.join(java_dir_full_path, file))
```
During the merging of the PR a small bug got introduced in those lines, I think this should now work and also be compatible with python 3.6

Baunsgaard replied:
This was me... there was a bug, i tried to fix, and apparently i did not finish the job.

kev-inn commented on src/main/python/systemds/__init__.py:
```diff
@@ -14,10 +14,4 @@
-__all__ = script.__all__
-__all__ += matrix.__all__
-__all__ += utils.__all__
+__all__ = ['matrix']
```
I tried to keep the public objects minimized for now, so restructuring the rest will be backwards compatible.

kev-inn commented on src/main/python/systemds/utils/helpers.py:
```diff
@@ -15,34 +15,16 @@
-AUTO_START_JMLC_SERVER = True
-
-
-def set_auto_start_jmlc(should_start_server: bool) -> None:
-    """
-    Can activate/deactivate automatic start of JMLC API Gateway.
-    Has to be called before any communication took place (`.compute()` was called).
-
-    :param should_start_server: activate/deactivate
-    """
-    global AUTO_START_JMLC_SERVER
-    AUTO_START_JMLC_SERVER = should_start_server
```
I removed this and switched to on demand auto starting of jmlc

Baunsgaard replied:
:+1:

kev-inn commented:
I will wait for the python test and check if it now works. Not sure yet.

kev-inn commented:
Oh and it would be nice if somebody could try building the python doc (read `src/main/python/docs/README.md` for instructions) and take a look. It is extremely rough around the edges, so any comments are appreciated.
Edit:
I would actually appreciate a design discussion on functions like `full`, `seq` and `federated` returning an `OperationNode`, but `class Matrix`, which is used for reading matrices from files or creating a matrix from numpy, being derived from `OperationNode`.

Baunsgaard commented on src/main/python/BUILD_INSTRUCTIONS.md:
```diff
@@ -0,0 +1,26 @@
+# Build instructions
+
+## Basic steps
```
newline after any # headers

Baunsgaard commented on src/main/python/BUILD_INSTRUCTIONS.md:
```diff
@@ -0,0 +1,26 @@
+
+## Basic steps
+The following steps have to be done for both cases
+- Build SystemDS with maven first `mvn package -DskipTests`, with the working directory being `SYSTEMDS_ROOT` (Root directory of SystemDS)
```
newline before any lists

Baunsgaard commented on src/main/python/BUILD_INSTRUCTIONS.md:
```diff
@@ -0,0 +1,26 @@
+# Build instructions
```
missing license

Baunsgaard commented on src/main/python/docs/Makefile:
```diff
@@ -0,0 +1,20 @@
+# Minimal makefile for Sphinx documentation
```
License

kev-inn replied:
Well this makefile was create by the `sphinx-quickstart` command, so I do not know how copyright will be handled here.

Baunsgaard commented on src/main/python/docs/README.md:
```diff
@@ -0,0 +1,9 @@
+# Building the docs
```
license

Baunsgaard commented on src/main/python/docs/make.bat:
```diff
@@ -0,0 +1,35 @@
+@ECHO OFF
```
license

kev-inn replied:
Same as with the `Makefile`

Baunsgaard commented on src/main/python/docs/requires-docs.txt:
```diff
@@ -0,0 +1,2 @@
+sphinx
```
if this file allows it ... license

Baunsgaard commented on src/main/python/docs/source/index.rst:
```diff
@@ -0,0 +1,33 @@
+SystemDS
```
license

Baunsgaard commented on src/main/python/docs/source/install.rst:
```diff
@@ -0,0 +1,28 @@
+Install SystemDS
```
license

Baunsgaard commented on src/main/python/docs/source/matrix.rst:
```diff
@@ -0,0 +1,39 @@
+Matrix API
```
license

Baunsgaard commented on src/main/python/docs/source/simple_examples.rst:
```diff
@@ -0,0 +1,77 @@
+Quickstart
```
license

Baunsgaard commented on src/main/python/systemds/matrix/operation_node.py:
```diff
@@ -14,23 +14,21 @@
 #  limitations under the License.
 # ------------------------------------------------------------------------------
 
-from __future__ import annotations
```
this was the line / import that broke 3.6 for me.

so now it works with 3.6?

kev-inn replied:
Yes :+1:

Baunsgaard commented on src/main/python/systemds/script_building/__init__.py:
```diff
@@ -0,0 +1,15 @@
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+# ------------------------------------------------------------------------------
```
Should the __init__ file be empty ?

kev-inn replied:
Yep, it can definitely be empty, and since we currently do not want anything public in this module it should be fine

Baunsgaard reviewed (commented):
:1st_place_medal: 
looks fine.
still missing the execution of the tests. for python 3.6 in the workflows

kev-inn commented:
Oh the rst comments did not work, I will fix it.

Baunsgaard committed:
[SYSTEMDS-313] Python Documentation Update via Github Actions

This commit contains updates to the python documentation and
minor changes to the way we generate the documentation artifacts using
github actions.

Also now when executing the build python prints to terminal also in git
actions.

Added dependency wheel for correctly building python API.

Minor changes to gitignore to ignore build artifacts from python.

Baunsgaard commented:
TODO:
Fix Python test in workflows, 
put the workflows that are currently disabled back.

sebwrede committed:
Merge branch 'master' into python_fixes
